adapter-pinecone/src/main/java/io/nosqlbench/adapter/pinecone/opdispensers/PineconeOpDispenser.java

```diff
@@ -73,6 +76,60 @@ protected ListValue generateListValue(List comparator) {
         return listValueBuilder.build();
     }
 
+    protected LongFunction<List<Float>> extractFloatVals(LongFunction<Object> af) {
+        return l -> this.getVectorValues(af.apply(l));
+    }
+
+    protected Map<String, Value> generateMetadataMap(Map<String, Object> metadata_values_map) {
+        Map<String, Value> metadata_map = new HashMap<>();
+        BiConsumer<String,Object> stringToValue = (key, val) -> {
+            Value targetval;
+            if (val instanceof String) targetval = Value.newBuilder().setStringValue((String)val).build();
+            else if (val instanceof Number) targetval = Value.newBuilder().setNumberValue((((Number) val).doubleValue())).build();
+            else if (val instanceof List) targetval = Value.newBuilder().setListValue(generateListValue((List) val)).build();
+            else if (val instanceof Boolean) targetval = Value.newBuilder().setBoolValue((Boolean) val).build();
+            else throw new RuntimeException("Unsupported metadata value type");
+            metadata_map.put(key, targetval);
+        };
+        metadata_values_map.forEach(stringToValue);
+        return metadata_map;
+    }
 
+    protected List<Float> getVectorValues(Object rawVectorValues) {
+        List<Float> floatValues;
+        if (rawVectorValues instanceof String) {
+            floatValues = new ArrayList<>();
+            String[] rawValues = (((String) rawVectorValues).split(","));
+            for (String val : rawValues) {
+                floatValues.add(Float.valueOf(val));
+            }
+        } else if (rawVectorValues instanceof List) {
+            floatValues = switch (((List<?>) rawVectorValues).get(0).getClass().getSimpleName()) {
+                case "Float" -> (List<Float>) rawVectorValues;
```

Comment:
Ok for now, but we should pre-qualify the type from a test value and make this not do type reflection on every call to the op dispenser. Once you know a given op template field produces a certain base value type, you can assume it continue to do the same. In other words, The instanceof checks should only be called in the op mapper or in the op dispenser constructor.